crates/circuit/src/dag_circuit.rs

```diff
@@ -1681,28 +1682,25 @@ def _format(operand):
         let py_op = op.extract::<OperationFromPython>()?;
         let qargs = qargs.map(|q| q.value);
         let cargs = cargs.map(|c| c.value);
-        let node = {
```

Comment:
We may want to avoid using the Rust `apply_operation_{back, front}` from within these Python versions, and keep using the internal `push_back` for them instead.

The primary reason being that the `qarg_indices` and `carg_indices` we have on lines 1687 and 1689 are _owned_. What we had before is a bit more efficient, since it moves those owned values directly into the `PackedInstruction`, and then moves that into `DAGCircuit::push_back`.

(FWIW, the new code here looks just about the same length as what we had before, so I don't think we'd be making much of a concession in terms of code simplicity.)

Reply:
I think this can be fixed easily considering that in most cases all the qargs/cargs will be owned as the user might construct this collection themselves. If we accepted an item that complies with the trait bounds `IntoIterator<Item = Qubit/Clbit>`, which would not discriminate the collection type as long as it has a `Bit` iterator type, it would technically achieve the same avoiding cloning the values manually.

However, I can also visualize how the user might instead want to re-use the Qargs/Cargs for something else down the line after insertion, in which case they would have to perform the copies manually. I say copies here since we've defined `Qargs/Cargs` as items that implement the `Copy` trait, which means that moving them should not be as costly.

The plan was to re-use the same methods so that they'd be tested without us having to build any new tests for them. But if a suggestion by one of your following comments (which asks to remove the insertion check from these rust native methods) is applied, we would still be required to build the PackedInstruction in this step which would require undoing the changes made.

Reply:
> I think this can be fixed easily considering that in most cases all the qargs/cargs will be owned as the user might construct this collection themselves. If we accepted an item that complies with the trait bounds IntoIterator<Item = Qubit/Clbit>, which would not discriminate the collection type as long as it has a Bit iterator type, it would technically achieve the same avoiding cloning the values manually.

I presume you're referring to the signature of the `DAGCircuit::apply_operation_back`, correct? The main issue with accepting a generic `IntoIterator` for `qargs` and `cargs` is that while it'd _allow_ an owned `Vec<{Qu|Cl}bit>` to be accepted, it wouldn't be possible to get it into the `Interner` without doing an allocation, since it is _more generic_ than the `Interner`'s `T::Owned` (which is exactly `Vec<T>`).

So you'd still need to collect those into `Vec<T>`, which would do an allocation.

For most use cases, it makes sense for the Rust-facing `DAGCircuit::apply_operation_back` to accept the bits as `&[{Qu,Cl}bit]`s (as you have it now) because more often than not, the `Interner` will get a cache hit during lookup, and it won't actually need to allocate. For `DAGCircuit::py_apply_operation_back` however, we already have an owned `Vec` from the Python conversion stuff, so we might as well ensure that we avoid the slim chance of that allocation happening by using the `Interner::insert_owned` methods directly. Hence why I think we ought to leave `DAGCircuit::py_apply_operation_{front,back}` as they are for now.

As @jakelishman mentioned, we might want to have some variants of `DAGCircuit::apply_operation_back` in the future that support scenarios where a Rust user already owns the bit lists, or where they pass `Interned` keys. If we do find a need for such methods, then perhaps at that point it makes sense to have `DAGCircuit::py_apply_operation_{front,back}` call those internally. But for now, `DAGCircuit::push_{front,back}` are already reasonable primitives for us to be using IMO in terms of code reuse.

> The plan was to re-use the same methods so that they'd be tested without us having to build any new tests for them. 

I like the pragmatics of this, but really we need to get Rust-based unit tests going for `DAGCircuit` sooner than later anyways. In the meantime, other Qiskit passes should exercise these at least in an integration testing capacity.